Better handling of images indexes

Michoko92 opened the pull request:
Filenames start with an index number, which is handy for file sorting by name. Until now, this index was computed by counting the number of files in the output directory + 1. However, if you made a generation, then deleted some files you were not interested in, then generated again, the index was starting again with a lower value than the last index used, mixing everything up.

This PR changes the way the index is computed, always determining the highest index present in the directory, then adding 1 to it.

Michoko92 committed:
Update images.py

Better computing of images indexes in filenames

AUTOMATIC1111 commented:
would fail with uppercase file extension. Do we even need to check for extension in the first place?

Does this work for grids?

Michoko92 commented:
Hi,

Checking for extensions is an extra check in case people output to a directory with mixed file types, but I suppose it's not necessary. I'll probably remove it to handle case inconsistencies.

About the grids, you're right, I think there will be an issue. I'll see if I can come up with a solution. Putting the PR into draft mode for now.

Michoko92 committed:
Update images.py

Handles grids names. Code more robust and doesn't fail if mixed with other files.

AUTOMATIC1111 commented:
the number is not necessarily at the end. You can, for example, setup filenames to have a date in them, and an image would be called 00005-2022-09-13.jpg

Michoko92 commented:
Please note that I make the following call in the code:

`basecount = get_next_sequence_number(path, basename != '')`

So when I call get_next_sequence_number, I force it to look at the start or the end of the filename, depending on the type of filename we are checking. In other words, if basename is empty, we are dealing with an image name, and I force get_next_sequence_number to look at the beginning of filenames. On the other hand, if basename is not empty, we are dealing with a grid for example, and we'll only check the sequence number at the end. It follows the same logic as this already existing code:

`fn = f"{basecount+i:05}" if basename == '' else f"{basename}-{basecount+i:04}"`

So in the example you gave above (with the date in the filename), it should work fine since the code already knows it has to check the sequence number at the beginning of the string.

AUTOMATIC1111 commented:
Well, an obvious outcome in the future would be for someone to request grids to have configurable names too, and that will be added, and then this would have to be fixed. Looking for the number at the end is just counter-productive when we know it's at start, either at very beginning, or after some non-numbers. Extract the digits from the last part of l[0] (using regex or whatever else) and that will be what we're looking for

Michoko92 commented:
When you say:

`Looking for the number at the end is just counter-productive when we know it's at start`

Well, I fully agree. That's where I don't understand what you mean, and that there may be a misunderstanding. Because that's exactly what I have been explaining with my two last messages: I DON'T look for the number at the end if it is at the start, as I always know where the number actually is, thanks to the checkAtEnd boolean parameter. I'm not trying to find a pattern, I know exactly where this sequence number is, as I just follow the same logic as what is already done in the "save_image" function:

`if basename is empty, the sequence number is at the beginning, else it's at the end`

So I can force get_next_sequence_number to look for the sequence number at the right position only (and I do).

We agree that sequence numbers will be always hardcoded at the start or the end of a filename, right? (else it would defeat their purpose, and could be mixed up with other numbers, like seeds). So if they remain hardcoded at the start or the end of filenames, even if names get more configurable in the future through tags, we can still rely on that rule. So if one day, grid names become configurable with tags, and you add a [date] tag for example, we'll have something like this:

`grid-2022-09-13-00005.jpg`

and the code of this PR will still handle this fine, as it will pick the last item of the string ( l[-1] ).

Please, I know you are a very productive man, looking at code all day long, and you may only have limited time. But if you read my messages and the code carefully, you'll see that it is future proof. The only way it could get broken is if, one day, we decide that sequence numbers position is not hardcoded any longer, and can be anywhere in the filename (but again it would defeat their sorting purpose). 

If you still think it's not future-proof, can you give me one example of a filename that this code would not handle with the current PR implementation, please? Thank you.

AUTOMATIC1111 commented:
The name of the grid will be `grid00005-2022-09-13.jpg`, it will be at the start same as images for sorting purposes.

Michoko92 commented:
OK thank you, I was not aware that this change was considered. Do you think that once this change is done, it would be hardcoded too? Something like this in the "save_image" function:

fn = f"{basecount+i:05}" if basename == '' else f"{basecount+i:04}-{basename}"

If it's the case, I think the get_next_sequence_number function will be easily adjustable at this time. If it uses a more complex approach (like using tags), then I don't know, as I suppose many cases could happen (like having seeds looking like a sequence number for example, and having to understand what is what).

So I'll let you decide if my current PR is good enough for now, and can be easily modified when the grid names change. I can even actually make that change regarding grid names starting with the sequence number, if you want me too. Else, I'll try later to reimplement this once things have evolved and I can make tests based on concrete code. 

Thank you and have a nice day! :)

Michoko92 commented:
Please ignore my last message, I understand what you mean, now. Sorry, for the dialogue of the deaf, I'll see what I can do to improve the PR.

AUTOMATIC1111 commented:
i was going to just accept and change myself later when I have time but if you still want to, sure

Michoko92 committed:
Update images.py

Better code

AUTOMATIC1111 commented:
well done
